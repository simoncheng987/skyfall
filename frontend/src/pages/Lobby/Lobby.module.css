.root {
  text-align: center;
}

.button {
  max-width: 300px;
  width: 100%;
  height: 70px;
  margin-top: 15px;
}

.defaultText {
  color: white;
  font-family: 'Poppins';
  font-weight: 900;
  text-shadow: 0px 2px 3px #1c1c1c;
}

.roomId {
  font-size: 55px;
  -webkit-text-stroke: 3px black;
}

.playerName {
  font-size: 30px;
  -webkit-text-stroke: 1px black;
  white-space: nowrap;
}

.player {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: -webkit-center;
  max-width: 150px;
}

.playersContainer {
  display: flex;
  flex-direction: row;
  gap: 500px;
  margin: 50px 0;
}
