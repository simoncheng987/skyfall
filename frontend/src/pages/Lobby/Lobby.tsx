import React, { useEffect, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import AvatarFrame from '../../components/AvatarFrame';
import Button from '../../components/Button';
import ErrorToast from '../../components/ErrorToast';
import Title from '../../components/Title';
import { useClient } from '../../context/ClientProvider';
import GameSafetyCheck from '../../utils/GameSafetyCheck';
import PageScaffold from '../PageScaffold';
import styles from './Lobby.module.css';

interface LobbyRoomInfo {
  lives: number;
  roomId: number;
}

export default function Lobby() {
  GameSafetyCheck();

  const { state } = useLocation();
  const navigate = useNavigate();

  const WAITING_PLAYER = 'Waiting for player...';
  const MAX_PLAYER = 2;
  const WORD_LIST_NAME = 'default';

  const imageUrls = ['host.png', 'guest.png'];

  const { client, setOpponent, isHost } = useClient();

  const [players, setPlayers] = useState<[string, string][]>([]);

  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  useEffect(() => {
    const timeout = setInterval(() => {
      setError(false);
    }, 2000);

    return () => {
      clearTimeout(timeout);
    };
  }, [error]);

  const startGame = () =>
    client?.emit('game:start', (state as LobbyRoomInfo).lives, WORD_LIST_NAME);

  client?.on('broadcast:player-joined', (mapStr) => {
    const map = JSON.parse(mapStr);
    Object.entries(map).forEach(([key, value]) => {
      if (key !== client.id) {
        setOpponent(value as string);
      }
    });

    setPlayers(Object.entries(map));
  });

  useEffect(() => {
    if (!state) {
      navigate('/create');
    } else {
      client?.on('game:start-fail', (msg) => {
        setErrorMessage(msg);
        setError(true);
      });

      client?.once('game:start-success', (lives: number) => {
        navigate('/game', { state: lives });
      });
    }

    return () => {
      client?.off('game:start-fail');
      client?.off('broadcast:player-joined');
    };
  }, []);

  return (
    <PageScaffold>
      <div className={styles.root}>
        <Title className={styles.title} fontSize="110px" text="Lobby" colorScheme="pink" />
        <p className={`${styles.roomId} ${styles.defaultText}`}>
          Room ID: {state && (state as LobbyRoomInfo).roomId}
        </p>
        <div className={styles.playersContainer}>
          {players.map((map, index) => (
            <div className={styles.player} key={map[0]}>
              <AvatarFrame imgSrc={imageUrls[index]} />
              <p className={`${styles.playerName} ${styles.defaultText}`}>{map[1]}</p>
            </div>
          ))}

          {players.length < MAX_PLAYER && (
            <div className={styles.player}>
              <AvatarFrame />
              <p className={`${styles.playerName} ${styles.defaultText}`}>{WAITING_PLAYER}</p>
            </div>
          )}
        </div>
        {isHost && players.length >= MAX_PLAYER && (
          <Button className={styles.button} text="Start" onClick={startGame} />
        )}
        {error && <ErrorToast className={styles.toast} text={errorMessage} />}
      </div>
    </PageScaffold>
  );
}
