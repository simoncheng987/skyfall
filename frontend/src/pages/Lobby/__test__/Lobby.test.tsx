import React from 'react';
import { MemoryRouter } from 'react-router-dom';
// import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ClientContextProvider from '../../../context/ClientProvider';
import Lobby from '../Lobby';

describe('Lobby', () => {
  it('snapshot test', () => {
    const tree = renderer
      .create(
        <ClientContextProvider>
          <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
            <Lobby />
          </MemoryRouter>
        </ClientContextProvider>,
      )
      .toTree();
    expect(tree).toMatchSnapshot();
  });
});
