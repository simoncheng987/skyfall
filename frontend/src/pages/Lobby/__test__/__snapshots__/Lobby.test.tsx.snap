// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lobby snapshot test 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
            "pathname": "/",
          },
        ]
      }
    >
      <Lobby />
    </MemoryRouter>,
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "children": <Lobby />,
      "initialEntries": Array [
        Object {
          "key": "testKey",
          "pathname": "/",
        },
      ],
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "basename": undefined,
        "children": <Lobby />,
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "navigationType": "POP",
        "navigator": Object {
          "action": "POP",
          "back": [Function],
          "block": [Function],
          "createHref": [Function],
          "forward": [Function],
          "go": [Function],
          "index": 0,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {},
        "rendered": Object {
          "instance": null,
          "nodeType": "component",
          "props": Object {
            "children": <div
              className="root"
            >
              <Title
                className="title"
                colorScheme="pink"
                fontSize="110px"
                text="Lobby"
              />
              <p
                className="roomId defaultText"
              >
                Room ID: 
              </p>
              <div
                className="playersContainer"
              >
                <div
                  className="player"
                >
                  <AvatarFrame
                    className=""
                    imgSrc=""
                  />
                  <p
                    className="playerName defaultText"
                  >
                    Waiting for player...
                  </p>
                </div>
              </div>
            </div>,
            "grassEnabled": true,
          },
          "rendered": Object {
            "instance": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="children"
                >
                  <div
                    className="error"
                  >
                    <ForwardRef(ExclamationCircleIcon)
                      width={35}
                    />
                    <p>
                      Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                    </p>
                  </div>
                </div>,
                <img
                  alt="grass background"
                  className="grassImage"
                  src="/assets/grass.png"
                />,
              ],
              "className": "container",
            },
            "rendered": Array [
              Object {
                "instance": null,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="error"
                  >
                    <ForwardRef(ExclamationCircleIcon)
                      width={35}
                    />
                    <p>
                      Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                    </p>
                  </div>,
                  "className": "children",
                },
                "rendered": Array [
                  Object {
                    "instance": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <ForwardRef(ExclamationCircleIcon)
                          width={35}
                        />,
                        <p>
                          Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                        </p>,
                      ],
                      "className": "error",
                    },
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "nodeType": "host",
                        "props": Object {
                          "aria-hidden": "true",
                          "children": <path
                            clipRule="evenodd"
                            d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
                            fillRule="evenodd"
                          />,
                          "fill": "currentColor",
                          "viewBox": "0 0 20 20",
                          "width": 35,
                          "xmlns": "http://www.w3.org/2000/svg",
                        },
                        "rendered": Array [
                          Object {
                            "instance": null,
                            "nodeType": "host",
                            "props": Object {
                              "clipRule": "evenodd",
                              "d": "M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z",
                              "fillRule": "evenodd",
                            },
                            "rendered": Array [],
                            "type": "path",
                          },
                        ],
                        "type": "svg",
                      },
                      Object {
                        "instance": null,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10",
                        },
                        "rendered": Array [
                          "Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10",
                        ],
                        "type": "p",
                      },
                    ],
                    "type": "div",
                  },
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "nodeType": "host",
                "props": Object {
                  "alt": "grass background",
                  "className": "grassImage",
                  "src": "/assets/grass.png",
                },
                "rendered": Array [],
                "type": "img",
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;
